### ISSUE TITLE
Progress tracking for kafka sinks

### ISSUE THREAD

**Issue by quodlibetor** [MaterializeInc/materialize#1442]
Currently we don't do any progress tracking for sinks, so when materialized is restarted all data from our ingests will be processed and re-sent to kafka. That means that anything downstream of us will get doubled/tripled etc data sizes, which is bad, depending on the use-case.

**Comment by cuongdo** [comment]
@ruchirK I'd like to bump the priority on this. #1572 and any other type of sink we want to implement requires resolving this issue.

**Comment by ruchirK** [comment]
I am looking at this rn :)

**Comment by ruchirK** [comment]
Frank and I discussed different implementation strategies for solving this and I will attempt to write down a summary of various proposed solutions so we can go from there. I'll continue to use Kafka sinks as the motivating driver but note that we talked through approaches that could generalize to different forms of sinks.

## Write out to a new Kafka queue after each restart

This approach is pretty self explanatory. We avoid writing duplicate data to the same Kafka topic  by writing to a new Kafka topic after each restart. We leave the "old" Kafka topic available for users to use, just with limited data and they can cut over to the new topic as they see fit

### Pros
* Simple to understand and implement

### Cons
* Requires user to let us create new Kafka topics on demand (not sure if thats already a requirement elsewhere)
* Requires user to take on a lot of the burden of handling restarts -- they have to figure out the new sink Kafka topic and start reading from it
* Requires user to accept rereading everything from scratch. This is marginally better than the status quo because at least now the user will have a clear starting point to start rereading everything from, but this still forces the user to somehow handle duplicates
* Forces someone (us or user) to garbage collect old and unused Kafka topics
* Duplicates data stored in Kafka topics

## Write out data in a more structured form

(Apologies in advance if I bungle any differential terminology or math)

Specifically, instead of writing out the trace of updates to the kafka topic as we see them, call the Consolidate operator on updates before they are sent to sink. This will give us a consistent (set of) lower and upper bounds for which the updates are valid, and as the name implies, will consolidate updates to the same record (i.e. (key, val, time) triple). The first part is the thing that's relevant to us. Basically, batches let us partition time and let us say after a batch has completed *we do have not written any records such that record r has time >= (any element in) upper*. This becomes an upper watermark / *frontier* for timestamps that have been published

So the algorithm for writing data out to a kafka sink is sketched out as:

1. Push the output trace through a Consolidate operator to get Batches of output records, where each Batch is a set of (key, val, time, diff) updates with at most one record per (key, val, time) triple and a Description consisting of lower bounds and upper bounds
2. Write the Batch to Kafka -- we need to write out the full (val, time, diff) triple for recovery reasons
3. Write the new frontier (the set of batch upper bounds) to Kafka / sqlite / something else that can persist and map sink dataflow -> frontier durably across restarts (we may also need to have dataflow ids for this)

Now, when we restart after having written data to Kafka, the first thing we will need to check is: "did we shutdown in the middle of writing a batch to Kafka and if so -- what are those records?".

Pausing here to elaborate on whats going on -- if we shutdown cleanly on a Batch boundary, everything we have written to the Kafka topic has time t < all elements in the frontier. In that case, we can use the frontier as our watermark for what to write to Kafka. If we want to write a message with ts < all elements in the frontier, then Consolidate guarantees that we've already inserted in Kafka and we can ignore that. Otherwise, we can go ahead and write it to Kafka / update frontiers as defined above.

On the other hand, if we shut down in the middle of writing a Batch to Kafka, then our notion of a frontier will be stale, and the Kafka topic will have some records with timestamps >= (some element in) the frontier. We need to a. figure out what those elements are and b. not send those again after we've restarted.

To accomplish a) -- We can scan the Kafka topic after we've restarted, looking for elements in advance of the frontier as defined above and add them to our "set of records that we have to be careful about" -- I'll call this the `retraction set`

To accomplish b) -- as we get Batches to write to Kafka, we need to

1. check if the Batch has any messages that match (val, time) from messages in our retraction set
2. if so, combine the diffs of those messages and if nonzero write to kafka
3. remove those messages from the retraction set

note that the way I've described this leaves us open to having elements in the retraction set that are not matched by any of the Batches we want to write to Kafka -- that's almost certainly a bug. Theres also a slightly different more algebraic way to think about this where we just fold in all messages from the retraction set that match [lower, upper) for Batches (sorry for being sketchy with notation)

Basically, we have an approach that we can reason about and that we feel reasonably confident should work

### Pros

* Takes away the user burden on having to deal with restarts

### Cons

* Much more complex
* Forces us to write `(row, timestamp, diff)` triples to Kafka and this is a format that likely few people will be able to easily use. Might force us into writing shims for other formats
* Forces us to call Consolidate before writing stuff out -- unclear how this impacts our perf
* Forces us to ingest the entire Kafka sink topic on startup (note that we can likely optimize this away by storing a (sink_id frontier, offset) mapping)

**Comment by ruchirK** [comment]
At this point I think I need a little bit more product clarity on

1. What kinds of tradeoffs are we looking for in sink behaviour? (is writing to a new topic / file / etc on restart acceptable in all/many/some cases? do we want to expose it as an option?)
2. What kind of output format(s) do we want to see used with sinks? Specifically, if we end up needing to support Debezium style diff envelopes -- how should we reconcile that against the (val, ts, diff) format we currently propose writing out?

@frankmcsherry @cuongdo @rjnn

**Comment by cuongdo** [comment]
To document an out-of-band discussion, we're going with the "Write out to a new Kafka queue after each restart" approach for now. We will revisit with more detailed customer requirements.

**Comment by ruchirK** [comment]
Two quick questions re: the "write to a new Kafka topic after restart"

1. Should we have this only triggered if the user specifies an option / keyword in the CREATE SINK command, or should this happen always? My vote is that this behaviour should be opt-in and I can't think of a good keyword for it but strawman (bad) proposal would be "NONCE"

2. Right now I'm appending topic creation timestamp to the user specified timestamp (could also plumb through materialize startup timestamp which has some nice UX benefits) -- We'll probably need a way to expose the actual topic names / other metadata for sinks -- would something like SHOW FULL SINKS be appropriate for that?

@cuongdo / anyone else with opinions

**Comment by cuongdo** [comment]
1. I think the default behavior should prioritize correctness, which means writing to a new topic upon restart. Is there a scenario in which the existing, incorrect behavior would be preferred?
2. cc @JLDLaughlin and @benesch for the opinions

cc @rjnn

**Comment by benesch** [comment]
> Two quick questions re: the "write to a new Kafka topic after restart"
> 
> 1. Should we have this only triggered if the user specifies an option / keyword in the CREATE SINK command, or should this happen always? My vote is that this behaviour should be opt-in and I can't think of a good keyword for it but strawman (bad) proposal would be "NONCE"

I'd rather limit the number of options where possible. This is the only correct way to use a sink at the moment, so I think we should force it on users for now.

> 2. Right now I'm appending topic creation timestamp to the user specified timestamp (could also plumb through materialize startup timestamp which has some nice UX benefits) -- We'll probably need a way to expose the actual topic names / other metadata for sinks -- would something like SHOW FULL SINKS be appropriate for that?

What's the user specified timestamp?

Anyway, I vote we plumb the Materialize startup timestamp through, along with a random u64 generated at boot, since it's possible that the startup timestamps could collide.

**Comment by ruchirK** [comment]
@benesch edited my original comment - it was a typo I meant "user specified topic name"

**Comment by ruchirK** [comment]
The only reason to avoid forcing it on users (at the moment) I guess is that they may not have us set up to auto create new topics?

**Comment by ruchirK** [comment]
But agreed and will do re: startup timestamp + random number on startup

Do you all think there's a need for "SHOW FULL SINKS" w/ the full actual topic name info as well?

**Comment by benesch** [comment]
> The only reason to avoid forcing it on users (at the moment) I guess is that they may not have us set up to auto create new topics?

We can use the Kafka admin API to create topics even if topic auto creation is disabled, if that's what you mean? In fact it occurs to me that we should be doing that no matter what, because we probably want to make these one partition topics so that we can guarantee order?

> Do you all think there's a need for "SHOW FULL SINKS" w/ the full actual topic name info as well?

👍 although I think we can just jam that information into the standard `SHOW SINKS` output because it's pretty important.

**Comment by ruchirK** [comment]
> We can use the Kafka admin API to create topics even if topic auto creation is disabled, if that's > what you mean? In fact it occurs to me that we should be doing that no matter what, because
> we probably want to make these one partition topics so that we can guarantee order?

👍
